File: python/tests/test_discovery.py
"""
Discovery Pipeline Tests (YAML → output)
=========================================
Tests that the discovery functions correctly read YAML metadata and return
consistent, accurate results. Uses shared YAML fixtures from
tests/fixtures/yaml/.

These tests verify Pipeline 2: YAML metadata → user-facing discovery output.
All tests run offline using pre-loaded test caches.

Test IDs: DISC-01 through DISC-18
"""

import pytest
import yaml
from pathlib import Path
from io import StringIO
from unittest.mock import patch

# ---------------------------------------------------------------------------
# Fixture paths
# ---------------------------------------------------------------------------
REPO_ROOT = Path(__file__).resolve().parents[2]
YAML_FIXTURES = REPO_ROOT / "tests" / "fixtures" / "yaml"


def _load_test_cache() -> dict:
    """Load the test indicator cache from YAML fixture."""
    path = YAML_FIXTURES / "unicef_indicators_metadata.yaml"
    with open(path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)


# ---------------------------------------------------------------------------
# Patch helper: override the indicator registry cache with test data
# ---------------------------------------------------------------------------

@pytest.fixture(autouse=True)
def patch_indicator_cache(monkeypatch):
    """Pre-populate indicator_registry cache with test YAML data."""
    import unicefdata.indicator_registry as reg

    test_cache = _load_test_cache()
    monkeypatch.setattr(reg, "_indicator_cache", test_cache)

    # Patch _ensure_cache_loaded to return test data without HTTP
    monkeypatch.setattr(reg, "_ensure_cache_loaded", lambda: test_cache)


# ===========================================================================
# DISC-01 to DISC-05: get_dataflow_for_indicator
# ===========================================================================

class TestGetDataflowForIndicator:
    """Test dataflow resolution from indicator codes."""

    def test_disc01_cme_indicator_resolves_to_cme(self):
        """DISC-01: CME_MRY0T4 resolves to CME dataflow."""
        from unicefdata.indicator_registry import get_dataflow_for_indicator
        result = get_dataflow_for_indicator("CME_MRY0T4")
        assert result == "CME"

    def test_disc02_nutrition_indicator_resolves(self):
        """DISC-02: NT_ANT_HAZ_NE2 resolves to NUTRITION dataflow."""
        from unicefdata.indicator_registry import get_dataflow_for_indicator
        result = get_dataflow_for_indicator("NT_ANT_HAZ_NE2")
        assert result == "NUTRITION"

    def test_disc03_immunisation_indicator_resolves(self):
        """DISC-03: IM_MCV1 resolves to IMMUNISATION dataflow."""
        from unicefdata.indicator_registry import get_dataflow_for_indicator
        result = get_dataflow_for_indicator("IM_MCV1")
        assert result == "IMMUNISATION"

    def test_disc04_unknown_indicator_returns_default(self):
        """DISC-04: Unknown indicator returns GLOBAL_DATAFLOW."""
        from unicefdata.indicator_registry import get_dataflow_for_indicator
        result = get_dataflow_for_indicator("NONEXISTENT_CODE")
        assert result == "GLOBAL_DATAFLOW"

    def test_disc05_known_override_bypasses_cache(self):
        """DISC-05: Known override indicators return corrected dataflow."""
        from unicefdata.indicator_registry import get_dataflow_for_indicator
        result = get_dataflow_for_indicator("PT_F_20-24_MRD_U18_TND")
        assert result == "PT_CM"


# ===========================================================================
# DISC-06 to DISC-09: get_indicator_info
# ===========================================================================

class TestGetIndicatorInfo:
    """Test indicator metadata retrieval."""

    def test_disc06_known_indicator_returns_dict(self):
        """DISC-06: get_indicator_info returns dict for known indicator."""
        from unicefdata.indicator_registry import get_indicator_info
        info = get_indicator_info("CME_MRY0T4")
        assert info is not None
        assert isinstance(info, dict)
        assert info["name"] == "Under-five mortality rate"

    def test_disc07_indicator_has_category(self):
        """DISC-07: Indicator info includes category field."""
        from unicefdata.indicator_registry import get_indicator_info
        info = get_indicator_info("CME_MRY0T4")
        assert info["category"] == "CME"

    def test_disc08_unknown_indicator_returns_none(self):
        """DISC-08: get_indicator_info returns None for unknown indicator."""
        from unicefdata.indicator_registry import get_indicator_info
        info = get_indicator_info("TOTALLY_FAKE_IND")
        assert info is None

    def test_disc09_indicator_info_has_description(self):
        """DISC-09: Indicator info includes description."""
        from unicefdata.indicator_registry import get_indicator_info
        info = get_indicator_info("NT_ANT_HAZ_NE2")
        assert "stunted" in info["description"].lower()


# ===========================================================================
# DISC-10 to DISC-13: list_indicators
# ===========================================================================

class TestListIndicators:
    """Test indicator listing and filtering."""

    def test_disc10_list_all_indicators(self):
        """DISC-10: list_indicators returns all 5 test indicators."""
        from unicefdata.indicator_registry import list_indicators
        result = list_indicators()
        assert len(result) == 5

    def test_disc11_filter_by_dataflow(self):
        """DISC-11: list_indicators(dataflow="CME") returns only CME indicators."""
        from unicefdata.indicator_registry import list_indicators
        result = list_indicators(dataflow="CME")
        assert len(result) == 2
        assert "CME_MRY0T4" in result
        assert "CME_MRY0" in result
        assert "IM_MCV1" not in result

    def test_disc12_filter_by_name(self):
        """DISC-12: list_indicators(name_contains="mortality") returns matches."""
        from unicefdata.indicator_registry import list_indicators
        result = list_indicators(name_contains="mortality")
        assert len(result) == 2
        assert "CME_MRY0T4" in result
        assert "CME_MRY0" in result

    def test_disc13_filter_by_dataflow_and_name(self):
        """DISC-13: Combined filters narrow results correctly."""
        from unicefdata.indicator_registry import list_indicators
        result = list_indicators(dataflow="IMMUNISATION", name_contains="measles")
        assert len(result) == 1
        assert "IM_MCV1" in result


# ===========================================================================
# DISC-14 to DISC-16: search_indicators (stdout)
# ===========================================================================

class TestSearchIndicators:
    """Test search_indicators output (prints to stdout)."""

    def test_disc14_search_produces_output(self, capsys):
        """DISC-14: search_indicators("mortality") prints results to stdout."""
        from unicefdata.indicator_registry import search_indicators
        search_indicators("mortality")
        captured = capsys.readouterr()
        assert "CME_MRY0T4" in captured.out
        assert "mortality" in captured.out.lower()

    def test_disc15_search_by_category(self, capsys):
        """DISC-15: search_indicators(category="IMMUNISATION") shows only IM indicators."""
        from unicefdata.indicator_registry import search_indicators
        search_indicators(category="IMMUNISATION")
        captured = capsys.readouterr()
        assert "IM_MCV1" in captured.out
        assert "IM_DTP3" in captured.out
        # CME indicator should not appear
        assert "CME_MRY0T4" not in captured.out

    def test_disc16_search_no_results(self, capsys):
        """DISC-16: search_indicators("xyznonexistent") prints no-results message."""
        from unicefdata.indicator_registry import search_indicators
        search_indicators("xyznonexistent")
        captured = capsys.readouterr()
        # Should print something (even if "no results found")
        assert len(captured.out) > 0


# ===========================================================================
# DISC-17 to DISC-18: Prefix-based inference
# ===========================================================================

class TestPrefixInference:
    """Test _infer_category prefix-based dataflow detection."""

    def test_disc17_prefix_inference_cme(self):
        """DISC-17: CME prefix infers CME category."""
        from unicefdata.indicator_registry import _infer_category
        assert _infer_category("CME_SOME_NEW_IND") == "CME"

    def test_disc18_prefix_inference_nutrition(self):
        """DISC-18: NT prefix infers NUTRITION category."""
        from unicefdata.indicator_registry import _infer_category
        result = _infer_category("NT_SOME_NEW_IND")
        assert result == "NUTRITION"

File: R/tests/testthat/test-discovery.R
# ===========================================================================
# Discovery Pipeline Tests (YAML -> output)
# ===========================================================================
# Tests that discovery functions correctly read YAML metadata and return
# consistent, accurate results. Uses shared YAML fixtures from
# tests/fixtures/yaml/.
#
# These tests verify Pipeline 2: YAML metadata -> user-facing discovery.
# All tests run offline using pre-loaded test YAML caches.
#
# Test IDs: DISC-01 through DISC-18
# ===========================================================================

library(testthat)

# ---------------------------------------------------------------------------
# Fixture paths and helpers
# ---------------------------------------------------------------------------
# helper-fixtures.R is auto-loaded by testthat from tests/testthat/
YAML_FIXTURES <- get_yaml_fixtures_dir()

load_test_indicators <- function() {
  yaml::read_yaml(file.path(YAML_FIXTURES, "unicef_indicators_metadata.yaml"))
}

load_test_dataflows <- function() {
  yaml::read_yaml(file.path(YAML_FIXTURES, "_unicefdata_dataflows.yaml"))
}

# ---------------------------------------------------------------------------
# Portable discovery functions (mirror indicator_registry.R logic)
# These work directly on YAML data without requiring the R package installed.
# ---------------------------------------------------------------------------

disc_get_dataflow_for_indicator <- function(indicator_code, indicators,
                                            default = "GLOBAL_DATAFLOW") {
  if (indicator_code %in% names(indicators)) {
    cat <- indicators[[indicator_code]]$category
    if (!is.null(cat) && nzchar(cat)) return(cat)
  }
  # Prefix inference
  prefix <- sub("_.*", "", indicator_code)
  prefix_map <- list(
    CME = "CME", NT = "NUTRITION", IM = "IMMUNISATION",
    ED = "EDUCATION", PT = "CHILD_PROTECTION", WS = "WASH_HOUSEHOLDS",
    HVA = "HIV_AIDS", DM = "DEMOGRAPHICS", ECD = "ECD",
    FT = "GENDER", PV = "CHLD_PVTY"
  )
  if (prefix %in% names(prefix_map)) return(prefix_map[[prefix]])
  default
}

disc_get_indicator_info <- function(indicator_code, indicators) {
  indicators[[indicator_code]]
}

disc_list_indicators <- function(indicators, dataflow = NULL, name_contains = NULL) {
  result <- indicators
  if (!is.null(dataflow)) {
    result <- Filter(function(x) identical(x$category, dataflow), result)
  }
  if (!is.null(name_contains)) {
    pattern <- tolower(name_contains)
    result <- Filter(function(x) grepl(pattern, tolower(x$name), fixed = TRUE), result)
  }
  result
}

disc_search_indicators <- function(indicators, query = NULL, category = NULL, limit = 50) {
  result <- indicators

  if (!is.null(category)) {
    result <- Filter(function(x) toupper(x$category) == toupper(category), result)
  }

  if (!is.null(query)) {
    q <- tolower(query)
    result <- Filter(function(x) {
      grepl(q, tolower(x$code), fixed = TRUE) ||
        grepl(q, tolower(x$name), fixed = TRUE) ||
        grepl(q, tolower(x$description %||% ""), fixed = TRUE)
    }, result)
  }

  if (limit > 0 && length(result) > limit) {
    result <- result[seq_len(limit)]
  }

  result
}


# ===========================================================================
# DISC-01 to DISC-05: get_dataflow_for_indicator
# ===========================================================================

test_that("DISC-01: CME_MRY0T4 resolves to CME dataflow", {
  indicators <- load_test_indicators()
  result <- disc_get_dataflow_for_indicator("CME_MRY0T4", indicators)
  expect_equal(result, "CME")
})

test_that("DISC-02: NT_ANT_HAZ_NE2 resolves to NUTRITION dataflow", {
  indicators <- load_test_indicators()
  result <- disc_get_dataflow_for_indicator("NT_ANT_HAZ_NE2", indicators)
  expect_equal(result, "NUTRITION")
})

test_that("DISC-03: IM_MCV1 resolves to IMMUNISATION dataflow", {
  indicators <- load_test_indicators()
  result <- disc_get_dataflow_for_indicator("IM_MCV1", indicators)
  expect_equal(result, "IMMUNISATION")
})

test_that("DISC-04: unknown indicator returns GLOBAL_DATAFLOW default", {
  indicators <- load_test_indicators()
  result <- disc_get_dataflow_for_indicator("NONEXISTENT_CODE", indicators)
  expect_equal(result, "GLOBAL_DATAFLOW")
})

test_that("DISC-05: prefix inference works for codes not in cache", {
  indicators <- load_test_indicators()
  # ED prefix not in test cache, but prefix map should resolve it
  result <- disc_get_dataflow_for_indicator("ED_SOME_NEW_IND", indicators)
  expect_equal(result, "EDUCATION")
})


# ===========================================================================
# DISC-06 to DISC-09: get_indicator_info
# ===========================================================================

test_that("DISC-06: known indicator returns list with name", {
  indicators <- load_test_indicators()
  info <- disc_get_indicator_info("CME_MRY0T4", indicators)
  expect_false(is.null(info))
  expect_equal(info$name, "Under-five mortality rate")
})

test_that("DISC-07: indicator info includes category", {
  indicators <- load_test_indicators()
  info <- disc_get_indicator_info("CME_MRY0T4", indicators)
  expect_equal(info$category, "CME")
})

test_that("DISC-08: unknown indicator returns NULL", {
  indicators <- load_test_indicators()
  info <- disc_get_indicator_info("TOTALLY_FAKE_IND", indicators)
  expect_null(info)
})

test_that("DISC-09: indicator info includes description", {
  indicators <- load_test_indicators()
  info <- disc_get_indicator_info("NT_ANT_HAZ_NE2", indicators)
  expect_true(grepl("stunted", info$description, ignore.case = TRUE))
})


# ===========================================================================
# DISC-10 to DISC-13: list_indicators
# ===========================================================================

test_that("DISC-10: list all indicators returns 5", {
  indicators <- load_test_indicators()
  result <- disc_list_indicators(indicators)
  expect_equal(length(result), 5)
})

test_that("DISC-11: filter by dataflow CME returns 2 indicators", {
  indicators <- load_test_indicators()
  result <- disc_list_indicators(indicators, dataflow = "CME")
  expect_equal(length(result), 2)
  expect_true("CME_MRY0T4" %in% names(result))
  expect_true("CME_MRY0" %in% names(result))
  expect_false("IM_MCV1" %in% names(result))
})

test_that("DISC-12: filter by name 'mortality' returns 2 matches", {
  indicators <- load_test_indicators()
  result <- disc_list_indicators(indicators, name_contains = "mortality")
  expect_equal(length(result), 2)
  expect_true("CME_MRY0T4" %in% names(result))
  expect_true("CME_MRY0" %in% names(result))
})

test_that("DISC-13: combined filters narrow results correctly", {
  indicators <- load_test_indicators()
  result <- disc_list_indicators(indicators, dataflow = "IMMUNISATION",
                                 name_contains = "measles")
  expect_equal(length(result), 1)
  expect_true("IM_MCV1" %in% names(result))
})


# ===========================================================================
# DISC-14 to DISC-16: search_indicators
# ===========================================================================

test_that("DISC-14: search 'mortality' finds CME indicators", {
  indicators <- load_test_indicators()
  result <- disc_search_indicators(indicators, query = "mortality")
  expect_true(length(result) >= 2)
  expect_true("CME_MRY0T4" %in% names(result))
})

test_that("DISC-15: search by category IMMUNISATION returns only IM indicators", {
  indicators <- load_test_indicators()
  result <- disc_search_indicators(indicators, category = "IMMUNISATION")
  expect_equal(length(result), 2)
  expect_true("IM_MCV1" %in% names(result))
  expect_true("IM_DTP3" %in% names(result))
  expect_false("CME_MRY0T4" %in% names(result))
})

test_that("DISC-16: search 'xyznonexistent' returns empty", {
  indicators <- load_test_indicators()
  result <- disc_search_indicators(indicators, query = "xyznonexistent")
  expect_equal(length(result), 0)
})


# ===========================================================================
# DISC-17 to DISC-18: Cross-format consistency
# ===========================================================================

test_that("DISC-17: indicator categories match dataflow IDs", {
  indicators <- load_test_indicators()
  dataflows <- load_test_dataflows()

  categories <- unique(vapply(indicators, function(x) x$category, character(1)))
  for (cat in categories) {
    expect_true(cat %in% names(dataflows),
                info = paste("Category", cat, "not found in dataflows YAML"))
  }
})

test_that("DISC-18: all CME indicators share same parent", {
  indicators <- load_test_indicators()
  cme_inds <- disc_list_indicators(indicators, dataflow = "CME")
  parents <- vapply(cme_inds, function(x) x$parent, character(1))
  expect_true(all(parents == "CME"))
})
